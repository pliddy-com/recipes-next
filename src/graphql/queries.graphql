# Query for Homepage RecipeCollection (all recipes)

query homePageQuery {
  recipeCollection(order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe collection based on filter

query recipeCollection($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    total
    items {
      ...RecipeDefault
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)

query listPageQuery($where: TagFilter) {
  tagCollection(where: $where, limit: 1) {
    total
    items {
      slug
      title
      linkedFrom {
        recipeCollection(limit: 24) {
          total
          items {
            ...RecipeSummary
          }
        }
      }
    }
  }
}

# Query for recipe slugs

query recipeSlugsCollection($where: RecipeFilter) {
  recipeCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query categorySlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query tagSlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
      linkedFrom {
        recipeCollection {
          total
        }
      }
    }
  }
}

# query for Taxonomy with Category & Subcategory levels

query taxonomyCollection($where: TaxonomyFilter) {
  taxonomyCollection(where: $where, limit: 1) {
    items {
      ...TaxonomyDefault

      tag {
        ...TagDefault
      }

      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            ...TagDefault

            linkedFrom {
              recipeCollection(limit: 1) {
                total
              }
            }
          }
          ...TaxonomyWithChildren
        }
      }
    }
  }
}
