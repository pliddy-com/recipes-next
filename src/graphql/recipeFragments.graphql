# Fragments with default fields for Recipe-specific sections:
#   * Ingredients
#   * Instructions
#   * Image
#   * Recipe

fragment IngredientsDefault on IngredientSection {
  sys {
    id
  }
  title
  slug
  label
  ingredientList
}

fragment InstructionsDefault on InstructionSection {
  sys {
    id
  }
  title
  slug
  label
  instructionList
}

fragment ImageDefault on Asset {
  sys {
    id
  }
  __typename
  title
  description
  contentType
  fileName
  size
  url
  height
  width
}

fragment DescriptionDefault on Recipe {
  description {
    json
    links {
      entries {
        inline {
          __typename
          sys {
            id
            __typename
          }
        }
        hyperlink {
          __typename
          sys {
            id
            __typename
          }
        }
        block {
          __typename

          sys {
            id
            __typename
          }
        }
      }
      assets {
        hyperlink {
          __typename
          title
        }
        block {
          __typename
          title
        }
      }
    }
  }
}

fragment RecipeDefault on Recipe {
  sys {
    id
  }
  __typename
  title
  slug
  description {
    json
  }
  abstract
  image {
    ...ImageDefault
  }
  ingredientsCollection(limit: 10) {
    items {
      ...IngredientsDefault
    }
  }
  equipment
  instructionsCollection(limit: 10) {
    items {
      ...InstructionsDefault
    }
  }
  notes
  tagsCollection(limit: 10) {
    items {
      ...TagDefault
    }
  }
}
