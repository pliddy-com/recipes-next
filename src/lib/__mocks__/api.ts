export interface ApiTestProps {
  getNavTaxonomy: () => [{ slug: string }];
  getRecipeSlugs: () => string[];
  getTagSlugs: () => string[];
  getRecipeIndex: () => [{ slug: string }];
  getRecipeList: () => [{ slug: string }];
  getRecipePage: () => { slug: string; title: string };
  getTagIndex: () => [{ slug: string; title: string }];
}

// import api library to mock
const api: ApiTestProps = jest.createMockFromModule('lib/api');

api.getNavTaxonomy = jest
  .fn()
  .mockResolvedValue([{ slug: 'slug-1', title: 'Title 1' }]);

api.getRecipeSlugs = jest.fn().mockResolvedValue(['slug-1', 'slug-2']);

api.getTagSlugs = jest.fn().mockResolvedValue(['slug-1', 'slug-2']);

api.getRecipeIndex = jest.fn().mockResolvedValue([
  {
    slug: 'slug-1',
    title: 'Title 1',
  },
  {
    slug: 'slug-2',
    title: 'Title 2',
  },
]);

api.getRecipeList = jest.fn().mockResolvedValue([
  {
    title: 'Recipe List Title',
    slug: 'recipe-list-title',
    linkedFrom: {
      recipeCollection: {
        items: [
          {
            slug: 'slug-1',
            title: 'Title 1',
          },
          {
            slug: 'slug-2',
            title: 'Title 2',
          },
        ],
      },
    },
  },
]);

api.getRecipePage = jest.fn().mockResolvedValue([
  {
    title: 'Recipe Title',
    slug: 'slug-1',
    abstract: 'abstract',
    image: {
      sys: {
        id: 'sysid-1',
        __typename: 'Sys',
      },
      __typename: 'Asset',
      title: 'Image Title',
      description: 'Image description.',
      contentType: 'image/jpeg',
      fileName: 'filename.jpg',
      size: 99999,
      url: 'https://recipes.pliddy.com/url',
      height: 300,
      width: 400,
    },
  },
]);

api.getTagIndex = jest.fn().mockResolvedValue([
  {
    __typename: 'Tag',
    title: 'Tag 1 Title',
    slug: 'tag-1',
    linkedFrom: {
      recipeCollection: {
        total: 2,
        items: [
          {
            __typename: 'Recipe',
            slug: 'slug-1',
            title: 'Title 2',
          },
          {
            __typename: 'Recipe',
            slug: 'slug-2',
            title: 'Title 1',
          },
        ],
      },
    },
  },
  {
    __typename: 'Tag',
    title: 'Tag 2 Title',
    slug: 'tag-2',
    linkedFrom: {
      recipeCollection: {
        total: 2,
        items: [
          {
            __typename: 'Recipe',
            slug: 'slug-3',
            title: 'Title 3',
          },
          {
            __typename: 'Recipe',
            slug: 'slug-4',
            title: 'Title 5',
          },
        ],
      },
    },
  },
]);

module.exports = api;
