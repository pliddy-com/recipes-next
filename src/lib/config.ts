interface PageProps {
  defaultDescription?: string;
  defaultTitle?: string;
  description?: string;
  domain?: string;
  title?: string;
}

interface MicrocopyProps {
  category?: PageProps;
  index?: PageProps;
  recipe?: PageProps;
  site?: PageProps;
  tag?: PageProps;
}

export interface ConfigProps {
  microcopy?: MicrocopyProps;
}

const microcopy: MicrocopyProps = {
  category: {
    defaultDescription: 'A collection of recipies categorized as',
    defaultTitle: 'Category',
  },
  index: {
    defaultTitle: 'All Recipes',
    description: "Patrick's recipe collection",
  },
  recipe: {
    defaultDescription: 'A recipe for',
    defaultTitle: 'Recipe',
  },
  site: {
    title: "Patrick's Recipes",
    domain: 'https://recipes.pliddy.com',
  },
  tag: {
    defaultDescription: 'A collection of recipies tagged as',
    defaultTitle: 'Tag',
  },
};

const config: ConfigProps = {
  microcopy,
};

export default config;
