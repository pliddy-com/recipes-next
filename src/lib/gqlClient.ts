import { TypedDocumentNode, createClient } from 'urql';
import { ResultOf, VariablesOf } from '@graphql-typed-document-node/core';

const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_ACCESS_TOKEN;
const API_ENDPOINT = process.env.NEXT_PUBLIC_API_ENDPOINT || '';

export const gqlClient = createClient({
  url: API_ENDPOINT,
  fetchOptions: () => ({
    headers: { authorization: ACCESS_TOKEN ? `Bearer ${ACCESS_TOKEN}` : '' },
  }),
});

export async function queryGraphQLContent<
  TDocument extends TypedDocumentNode<
    ResultOf<TDocument>,
    VariablesOf<TDocument>
  >,
  TVars = TDocument extends unknown ? never : VariablesOf<TDocument>
>(document: TDocument, variables: TVars) {
  const { data, error } = await gqlClient
    // @ts-expect-error query variables are typed by graphQLRequest
    .query(document, variables)
    .toPromise();

  if (!data) {
    if (error) {
      throw error;
    }
    throw new Error('No data returned');
  }

  return data;
}
