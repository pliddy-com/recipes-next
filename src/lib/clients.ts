import { TypedDocumentNode, createClient } from 'urql';
import { print } from 'graphql';
import { ResultOf, VariablesOf } from '@graphql-typed-document-node/core';

const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_ACCESS_TOKEN;
const API_ENDPOINT = process.env.NEXT_PUBLIC_API_ENDPOINT || '';

const gqlClient = createClient({
  url: API_ENDPOINT,
  fetchOptions: () => ({
    headers: { authorization: ACCESS_TOKEN ? `Bearer ${ACCESS_TOKEN}` : '' },
  }),
});

export async function queryGraphQLContent<
  TDocument extends TypedDocumentNode<
    ResultOf<TDocument>,
    VariablesOf<TDocument>
  >,
  TVars = TDocument extends unknown ? never : VariablesOf<TDocument>
>(document: TDocument, variables: TVars) {
  const { data, error } = await gqlClient
    // @ts-expect-error query variables are typed by graphQLRequest
    .query(document, variables)
    .toPromise();

  if (!data) {
    if (error) {
      throw error;
    }
    throw new Error('No data returned');
  }

  return data;
}

// use fetch for nav query instead of the gql client to minimize browser bundle size

export async function fetchContent<
  TDocument extends TypedDocumentNode<
    ResultOf<TDocument>,
    VariablesOf<TDocument>
  >,
  TVars = TDocument extends unknown ? never : VariablesOf<TDocument>
>(document: TDocument, variables: TVars) {
  try {
    const response = await window.fetch(API_ENDPOINT, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: ACCESS_TOKEN ? `Bearer ${ACCESS_TOKEN}` : '',
      },
      // send the GraphQL query
      body: JSON.stringify({ query: print(document), variables }),
    });

    const { data } = await response.json();

    return data;
  } catch (e) {
    console.error(e);
  }
}
