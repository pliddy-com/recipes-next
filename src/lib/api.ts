import { queryGraphQLContent } from 'lib/gqlClient';

import {
  NavTaxonomyDocument,
  NavTaxonomyQueryVariables,
  RecipeIndexDocument,
  RecipeIndexQueryVariables,
  RecipeListDocument,
  RecipeListQueryVariables,
  RecipePageDocument,
  RecipePageQueryVariables,
  RecipeSlugsDocument,
  RecipeSlugsQueryVariables,
  TagIndexDocument,
  TagIndexQueryVariables,
  TagSlugsDocument,
  TagSlugsQueryVariables,
} from 'types/queries';

import { hasValue } from 'lib/typeUtils';

// used to query content for nav menu taxonomy on client layout
export const getNavTaxonomy = async (variables: NavTaxonomyQueryVariables) => {
  const { taxonomyCollection } = await queryGraphQLContent(
    NavTaxonomyDocument,
    variables
  );

  return taxonomyCollection ? taxonomyCollection.items.filter(hasValue) : [];
};

// used by getStaticProps for recipe page
export const getRecipeSlugs = async (variables: RecipeSlugsQueryVariables) => {
  const { recipeCollection } = await queryGraphQLContent(
    RecipeSlugsDocument,
    variables
  );

  const results = recipeCollection?.items?.map((child) =>
    child && child.slug ? child.slug : null
  );

  return results ? results?.filter(hasValue) : [];
};

// used by getStaticProps for tag page
export const getTagSlugs = async (variables: TagSlugsQueryVariables) => {
  const { tagCollection } = await queryGraphQLContent(
    TagSlugsDocument,
    variables
  );

  const results = tagCollection?.items
    .filter(hasValue)
    .filter(
      (item) =>
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0
    )
    .map((child) => (child && child.slug ? child.slug : null));

  return results ? results?.filter(hasValue) : [];
};

// used to query content for home index page
export const getRecipeIndex = async (variables?: RecipeIndexQueryVariables) => {
  const { recipeCollection } = await queryGraphQLContent(
    RecipeIndexDocument,
    variables
  );

  return recipeCollection ? recipeCollection.items.filter(hasValue) : [];
};

// used to query content for tag & category pagepage
export const getRecipeList = async (variables: RecipeListQueryVariables) => {
  const { tagCollection } = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  return tagCollection ? tagCollection.items.filter(hasValue) : [];
};

// used to query content for standalone recipe page
export const getRecipePage = async (variables?: RecipePageQueryVariables) => {
  const { recipeCollection } = await queryGraphQLContent(
    RecipePageDocument,
    variables
  );

  return recipeCollection ? recipeCollection.items.filter(hasValue) : [];
};

// used to query content for tag index page
export const getTagIndex = async (variables?: TagIndexQueryVariables) => {
  const { tagCollection } = await queryGraphQLContent(
    TagIndexDocument,
    variables
  );

  const results = tagCollection?.items
    // .filter(hasValue)
    .filter(
      (item) =>
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0
    );
  // .map((child) => (child && child.slug ? child.slug : null));

  return results ? results?.filter(hasValue) : [];
};
