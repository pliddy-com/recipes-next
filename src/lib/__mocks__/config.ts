import { ConfigProps } from 'lib/config';

const microcopy = {
  category: {
    defaultDescription: 'Category default description',
    defaultTitle: 'Category Default Title',
  },
  index: {
    defaultTitle: 'Index Default Title',
    description: 'Index default description.',
  },
  recipe: {
    defaultDescription: 'Recipe default description',
    defaultTitle: 'Recipe Default Title',
  },
  site: {
    title: 'Site Title',
  },
  tag: {
    defaultDescription: 'Tag default description',
    defaultTitle: 'Tag Default Title',
  },
};

const config: ConfigProps = {
  microcopy,
};

module.exports = config;
