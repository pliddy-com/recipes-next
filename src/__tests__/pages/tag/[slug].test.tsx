// import testing-library methods
import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

// import the component to test
import TagPage, { getStaticPaths, getStaticProps } from 'pages/tag/[slug]';
import { ListPageItemFragment } from 'types/generated/graphql';

import * as api from 'lib/api';
import config from 'lib/config';

jest.mock('lib/api');
jest.mock('lib/config');
jest.mock('components/PageHeadTag/PageHeadTag');

const pageContentData = { content: 'list page content' };
const tagSlugData = { slug: 'slug-1' };

describe('TagPage in tag/[slug].tsx', () => {
  // reset mocks after each test
  afterEach(() => {
    jest.resetModules();
  });

  describe('when there is page content', () => {
    const pageContent = {
      slug: 'baking',
      title: 'Baking',
      linkedFrom: {
        recipeCollection: {
          total: 3,
          items: [
            {
              sys: {
                id: 'sysid-0',
                __typename: 'Sys',
              },
              __typename: 'Recipe',
              title: 'Recipe 1 Title',
              slug: 'recipe-1-title',
              abstract: 'Recipe 1 abstract.',
              image: {
                sys: {
                  id: 'sysid-1',
                  __typename: 'Sys',
                },
                __typename: 'Asset',
                title: 'Image 1 Title',
                description: 'Image 1 description',
                contentType: 'image/jpeg',
                fileName: 'image1.jpg',
                size: 99999,
                url: 'https://test.url/biscuits.jpg',
                height: 300,
                width: 400,
              },
              tagsCollection: {
                items: [
                  {
                    sys: {
                      id: 'sysid-2',
                      __typename: 'Sys',
                    },
                    __typename: 'Tag',
                    title: 'Tag 1',
                    slug: 'tag-1',
                  },
                  {
                    sys: {
                      id: 'sysid-3',
                      __typename: 'Sys',
                    },
                    __typename: 'Tag',
                    title: 'Tag 2',
                    slug: 'tag-2',
                  },
                ],
                __typename: 'RecipeTagsCollection',
              },
            },
          ],
          __typename: 'RecipeCollection',
        },
        __typename: 'TagLinkingCollections',
      },
      __typename: 'Tag',
    };

    it('it renders the page', async () => {
      const propsContext = {
        preview: false,
        params: tagSlugData,
      };

      const expectedProps = {
        props: { pageContent: pageContentData, preview: false },
      };

      const expectedPaths = {
        fallback: false,
        paths: [{ params: tagSlugData }],
      };

      const queryTagSlugsSpy = jest.spyOn(api, 'queryTagSlugs');
      const queryListPageContentSpy = jest.spyOn(api, 'queryListPageContent');

      const { container } = render(
        <TagPage
          pageContent={pageContent as ListPageItemFragment}
          preview={false}
        />
      );

      expect(await getStaticProps(propsContext)).toEqual(expectedProps);
      expect(await getStaticPaths({})).toEqual(expectedPaths);

      expect(queryTagSlugsSpy).toHaveBeenCalled();
      expect(queryListPageContentSpy).toHaveBeenCalled();

      // assert that the component matches the existing snapshot
      expect(container).toMatchSnapshot();

      // assert that page head tags are rendered
      const titleTag = document.getElementsByTagName('title')[0];
      expect(titleTag).toBeInTheDocument();
      expect(titleTag.text.includes(pageContent.title));

      // assert that page container is rendered
      const page = document.querySelector('.page');
      expect(page).toBeInTheDocument();
    });

    describe('when the slug is not a string', () => {
      it('it does not render the page', async () => {
        const propsContext = {
          preview: false,
          params: { slug: undefined },
        };

        try {
          await getStaticProps(propsContext);
        } catch (e) {
          expect(e).toEqual(new Error('Error in SSG!'));
        }
      });
    });
  });

  describe('when there is no page content', () => {
    // before each test, delete microcopy node from config
    beforeEach(() => {
      delete config.microcopy;
    });

    it('it does not render the category page', () => {
      const pageContent = undefined as unknown as ListPageItemFragment;

      render(<TagPage pageContent={pageContent} preview={false} />);

      // test if card compoent is not rendered
      const page = document.querySelector('.page');
      expect(page).toBeNull();
    });
  });
});
