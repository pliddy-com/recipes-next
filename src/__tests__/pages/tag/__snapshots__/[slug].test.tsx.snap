// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagPage in tag/[slug].tsx when there is page content it renders the page 1`] = `
<div>
  <title>
    Baking Tag Default Title
  </title>
  <meta
    content="Tag default description Baking"
    name="description"
  />
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg page css-1oqqzyl-MuiContainer-root"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
    >
      Baking Tag Default Title
    </h1>
    <h2
      class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-1wniyei-MuiTypography-root"
    >
      1 Total
    </h2>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 css-wq3jbm-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root recipe css-1v5z4dq-MuiPaper-root-MuiCard-root"
        >
          <a
            class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
            href="/recipe/recipe-1-title"
            tabindex="0"
          >
            <div
              class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
                >
                  Recipe 1 Title
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiCardHeader-subheader css-nrdprl-MuiTypography-root"
                >
                  Tag 1
                </span>
              </div>
            </div>
            <div
              class="MuiCardMedia-root css-16cvb1l-MuiCardMedia-root"
            >
              <div
                class="imageWrapper MuiBox-root css-0"
              >
                <link
                  as="image"
                  href="https://test.url/biscuits.jpg"
                  imagesizes="450px"
                  imagesrcset="https://test.url/biscuits.jpg?w=450&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=900&fm=webp&q=75 2x"
                  media="(min-width: 669px)"
                  rel="preload"
                />
                <link
                  as="image"
                  href="https://test.url/biscuits.jpg"
                  imagesizes="300px"
                  imagesrcset="https://test.url/biscuits.jpg?w=300&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=600&fm=webp&q=75 2x"
                  media="(min-width: 600px) and (max-width: 668px)"
                  rel="preload"
                />
                <link
                  as="image"
                  href="https://test.url/biscuits.jpg"
                  imagesizes="600px"
                  imagesrcset="https://test.url/biscuits.jpg?w=600&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=1200&fm=webp&q=75 2x"
                  media="(min-width: 485px) and (max-width: 599px)"
                  rel="preload"
                />
                <link
                  as="image"
                  href="https://test.url/biscuits.jpg"
                  imagesizes="450px"
                  imagesrcset="https://test.url/biscuits.jpg?w=450&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=900&fm=webp&q=75 2x"
                  media="(min-width: 335px) and (max-width: 484px)"
                  rel="preload"
                />
                <link
                  as="image"
                  href="https://test.url/biscuits.jpg"
                  imagesizes="300px"
                  imagesrcset="https://test.url/biscuits.jpg?w=300&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=600&fm=webp&q=75 2x"
                  media="(max-width: 334px)"
                  rel="preload"
                />
                <picture>
                  <source
                    media="(min-width: 669px)"
                    sizes="450px"
                    srcset="https://test.url/biscuits.jpg?w=450&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=900&fm=webp&q=75 2x"
                  />
                  <source
                    media="(min-width: 600px) and (max-width: 668px)"
                    sizes="300px"
                    srcset="https://test.url/biscuits.jpg?w=300&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=600&fm=webp&q=75 2x"
                  />
                  <source
                    media="(min-width: 485px) and (max-width: 599px)"
                    sizes="600px"
                    srcset="https://test.url/biscuits.jpg?w=600&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=1200&fm=webp&q=75 2x"
                  />
                  <source
                    media="(min-width: 335px) and (max-width: 484px)"
                    sizes="450px"
                    srcset="https://test.url/biscuits.jpg?w=450&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=900&fm=webp&q=75 2x"
                  />
                  <source
                    media="(max-width: 334px)"
                    sizes="300px"
                    srcset="https://test.url/biscuits.jpg?w=300&fm=webp&q=75 1x, https://test.url/biscuits.jpg?w=600&fm=webp&q=75 2x"
                  />
                  <img
                    alt="Image 1 description"
                    height="300"
                    loading="eager"
                    src="https://test.url/biscuits.jpg"
                    style="max-width: 100%; height: auto;"
                    width="400"
                  />
                </picture>
              </div>
            </div>
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 abstract css-e784if-MuiTypography-root"
              >
                Recipe 1 abstract.
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <div
              class="css-180l46g-MuiStack-root"
              data-testid="tag-buttons"
            >
              <a
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-wmndcn-MuiButtonBase-root-MuiButton-root"
                href="/tag/tag-1"
                role="button"
                tabindex="0"
              >
                 Tag 1 
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </a>
              <a
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-wmndcn-MuiButtonBase-root-MuiButton-root"
                href="/tag/tag-2"
                role="button"
                tabindex="0"
              >
                 Tag 2 
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
